readme.md
# BitSet

A simple bit set implementation for compact storage of sets of values. The library itself is stateless providing only pure mapping functions.

## Requirements

PHP 7.1+

## Installation

    $ composer require pwm/bitset

## Usage

In this example we are going to use a bit set to put users in groups.

Let's create a `Group` class. It has a list of groups, named G1 to G5 for this example, and a map between groups and binary values of powers of 2 provided by BitSet. Finally it has 2 functions that map between group names and bit values.

```php
class Group
{
    const G1 = 'Group 1';
    const G2 = 'Group 2';
    const G3 = 'Group 3';
    const G4 = 'Group 4';
    const G5 = 'Group 5';

    private const MAP = [
        self::G1 => BitSet::B1,
        self::G2 => BitSet::B2,
        self::G3 => BitSet::B3,
        self::G4 => BitSet::B4,
        self::G5 => BitSet::B5,
    ];

    public static function fromGroups(array $groups): array
    {
        return array_values(array_intersect_key(self::MAP, array_flip($groups)));
    }

    public static function toGroups(array $bitValues): array
    {
        return array_values(array_intersect_key(array_flip(self::MAP), array_flip($bitValues)));
    }
}
```

Let's also create a `User` class. A user can belong to groups. The `$groups` property starts from 0 that represents the empty set. It has 5 methods to manipulate its groups, corresponding to BitSet's 5 functions: `get()`, `set()`, `add()`, `remove()` and `has()`.

```php
class User
{
    private $groups = BitSet::EMPTY;

    public function getGroups(): array
    {
        return Group::toGroups(BitSet::get($this->groups));
    }

    public function setGroups(array $groups): void
    {
        $this->groups = BitSet::set(Group::fromGroups($groups));
    }

    public function addGroups(array $groups): void
    {
        $this->groups = BitSet::add($this->groups, Group::fromGroups($groups));
    }

    public function removeGroups(array $groups): void
    {
        $this->groups = BitSet::remove($this->groups, Group::fromGroups($groups));
    }

    public function hasGroup(string $group): bool
    {
        $a = Group::fromGroups([$group]);
        return isset($a[0])
            ? BitSet::has($this->groups, $a[0])
            : false;
    }
}
```

Now we can use the above the following way:

```php
$user = new User();
$user->setGroups([Group::G1, Group::G2]);

assert($user->getGroups() === [Group::G1, Group::G2]);
assert($user->hasGroup(Group::G1) === true);

$user->addGroups([Group::G4, Group::G5]);
$user->removeGroups([Group::G1, Group::G4]);

assert($user->getGroups() === [Group::G2, Group::G5]);
assert($user->hasGroup(Group::G1) === false);
```

## How it works

As an example let's take the numbers 1, 4 and 8. Their sum is 13. Pretty uninteresting so far. But let's look at their binary representations: 0b1, 0b100 and 0b1000 respectively. Their sum 13 is 0b1101. See how the base-2 representations of 1, 4 and 8 "fill unique slots" in 13's, leaving only one slot unset? This is because 1, 4 and 8 are all powers of 2 so their binary representations starts with a 1 followed by zero or more 0s. 0b1 = 2^0 = 1, 0b100 = 2^2 = 4, 0b1000 = 2^3 = 8. Their sum: 2^0 + 2^2 + 2^3 = 0b1 + 0b100 + 0b1000 = 0b1101 = 13. Now, say in a game, we could call 1 "north", 2 "south", 4 "east" and 8 "west" and then 13 could mean that our character can move all directions from its current position except south as 2 (0b10) is not in 13 (0b1101).

The idea behind a bit set is using a base-2 representation of a number to singal the presence or absence of values by having its bits set to 1 or 0. The values themselves are also numbers with the restriction that they must be powers of 2 starting from 2^0 and in general, depending on the underlying system, going up to 2^32 or 2^64. For example on a 64 bit system one can store up to 64 unique values in a bit set.

The `BitSet` library itself is a collection of 5 pure functions that map between values and their bit set:

`set(array $values): int`

Maps a list of unique positive integers to a single integer, the bit set, that is their sum. All integers in the input list must be powers of 2.

`get(int $bitSet): array`

Maps an integer, the bit set, to a list of unique positive integers. All integers in the output list must be powers of 2.

`add(int $bitSet, array $values): int`

Adds a list of unique positive integers to a bit set. All integers in the list must be powers of 2.

`remove(int $bitSet, array $values): int`

Subtracts a list of unique positive integers from a bit set. All integers in the list must be powers of 2.

`has(int $bitSet, int $value): bool`

Predicate that tells whether an element is in a bit set or not.

## Tests

	$ vendor/bin/phpunit

## Changelog

[Click here](changelog.md)


changelog.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2017-07-22
### Added
  * Bit constants
  
## [1.0.0] - 2017-07-09
### Added
  * Initial release


tests/unit/BitSetTest.php
<?php
declare(strict_types=1);

namespace Pwm\BitSet;

use PHPUnit\Framework\TestCase;

class BitSetTest extends TestCase
{
    /**
     * @test
     */
    public function get_values_from_bit_set(): void
    {
        static::assertSame([], BitSet::get(0b0));
        static::assertSame([0b1], BitSet::get(0b1));
        static::assertSame([0b1, 0b10], BitSet::get(0b11));
        static::assertSame([0b1, 0b100], BitSet::get(0b101));
        static::assertSame([0b1, 0b10, 0b100], BitSet::get(0b111));
    }

    /**
     * @test
     */
    public function set_bit_set_from_values(): void
    {
        static::assertSame(0b0, BitSet::set([]));
        static::assertSame(0b1, BitSet::set([0b1]));
        static::assertSame(0b11, BitSet::set([0b1, 0b10]));
        static::assertSame(0b101, BitSet::set([0b1, 0b100]));
        static::assertSame(0b111, BitSet::set([0b1, 0b10, 0b100]));
    }

    /**
     * @test
     * @expectedException \InvalidArgumentException
     * @dataProvider provideMultiValuedBitArrays
     */
    public function set_does_not_work_with_multi_bit_values(array $multiBitValues): void
    {
        BitSet::set($multiBitValues);
    }

    /**
     * @test
     */
    public function set_value_order_does_not_matter(): void
    {
        static::assertSame(BitSet::set([0b1, 0b10, 0b100]), BitSet::set([0b100, 0b1, 0b10]));
    }

    /**
     * @test
     */
    public function add_values_to_bit_set(): void
    {
        static::assertSame(0b1, BitSet::add(0b0, [0b1]));
        static::assertSame(0b11, BitSet::add(0b1, [0b10]));
        static::assertSame(0b101, BitSet::add(0b1, [0b100]));
        static::assertSame(0b111, BitSet::add(0b0, [0b1, 0b10, 0b100]));
        static::assertSame(0b1111, BitSet::add(0b1010, [0b1, 0b100]));
    }

    /**
     * @test
     * @expectedException \InvalidArgumentException
     * @dataProvider provideMultiValuedBitArrays
     */
    public function add_does_not_work_with_multi_bit_values(array $multiBitValues): void
    {
        BitSet::add(0b0, $multiBitValues);
    }

    /**
     * @test
     */
    public function add_value_order_does_not_matter(): void
    {
        static::assertSame(BitSet::add(0b0, [0b1, 0b10, 0b100]), BitSet::add(0b0, [0b100, 0b1, 0b10]));
    }

    /**
     * @test
     */
    public function adding_zero_has_no_effect(): void
    {
        static::assertSame(0b1, BitSet::add(0b1, [0b0]));
        static::assertSame(BitSet::add(0b10, [0b1]), BitSet::add(0b10, [0b0, 0b1]));
    }

    /**
     * @test
     */
    public function add_is_idempotent(): void
    {
        static::assertSame(BitSet::add(0b0, [0b1]), BitSet::add(BitSet::add(0b0, [0b1]), [0b1]));
    }

    /**
     * @test
     */
    public function remove_values_from_bit_set(): void
    {
        static::assertSame(0b0, BitSet::remove(0b1, [0b1]));
        static::assertSame(0b1, BitSet::remove(0b11, [0b10]));
        static::assertSame(0b10, BitSet::remove(0b11, [0b1]));
        static::assertSame(0b0, BitSet::remove(0b111, [0b1, 0b10, 0b100]));
        static::assertSame(0b1010, BitSet::remove(0b1111, [0b1, 0b100]));
    }

    /**
     * @test
     * @expectedException \InvalidArgumentException
     * @dataProvider provideMultiValuedBitArrays
     */
    public function remove_does_not_work_with_multi_bit_values(array $multiBitValues): void
    {
        BitSet::remove(0b111, $multiBitValues);
    }

    /**
     * @test
     */
    public function remove_value_order_does_not_matter(): void
    {
        static::assertSame(BitSet::remove(0b111, [0b1, 0b10, 0b100]), BitSet::remove(0b111, [0b100, 0b1, 0b10]));
    }

    /**
     * @test
     */
    public function removing_zero_has_no_effect(): void
    {
        static::assertSame(0b1, BitSet::remove(0b1, [0b0]));
        static::assertSame(BitSet::remove(0b11, [0b1]), BitSet::remove(0b11, [0b0, 0b1]));
    }

    /**
     * @test
     */
    public function remove_is_idempotent(): void
    {
        static::assertSame(BitSet::remove(0b111, [0b10]), BitSet::remove(BitSet::remove(0b111, [0b10]), [0b10]));
    }

    /**
     * @test
     */
    public function has_checks_if_the_given_values_are_elements_of_the_bit_set(): void
    {
        static::assertTrue(BitSet::has(0b1, 0b1));
        static::assertFalse(BitSet::has(0b1, 0b10));

        static::assertTrue(BitSet::has(0b11, 0b1));
        static::assertTrue(BitSet::has(0b11, 0b10));
        static::assertFalse(BitSet::has(0b11, 0b100));

        static::assertTrue(BitSet::has(0b100, 0b100));
        static::assertFalse(BitSet::has(0b100, 0b1));
        static::assertFalse(BitSet::has(0b100, 0b10));
    }

    /**
     * @test
     * @expectedException \InvalidArgumentException
     * @dataProvider provideMultiValuedBits
     */
    public function has_does_not_work_with_multi_bit_values(int $multiBitValue): void
    {
        BitSet::has(0b111, $multiBitValue);
    }

    /**
     * @test
     */
    public function zero_is_always_an_element(): void
    {
        static::assertTrue(BitSet::has(0b0, 0b0));
        static::assertTrue(BitSet::has(0b1, 0b0));
    }

    public function provideMultiValuedBitArrays(): array
    {
        return [
            [[0b11]],
            [[0b101]],
            [[0b1, 0b110]],
        ];
    }

    public function provideMultiValuedBits(): array
    {
        return [
            [0b11],
            [0b101],
        ];
    }
}


tests/unit/BitSet32BitTest.php
<?php
declare(strict_types=1);

namespace Pwm\BitSet;

use PHPUnit\Framework\TestCase;

class BitSet32BitTest extends TestCase
{
    /**
     * @test
     */
    public function setting_all_32_bits_result_in_the_max_signed_integer_on_32_bit_systems(): void
    {
        $b32a = [
            BitSet::B0,  BitSet::B1,  BitSet::B2,  BitSet::B3,
            BitSet::B4,  BitSet::B5,  BitSet::B6,  BitSet::B7,
            BitSet::B8,  BitSet::B9,  BitSet::B10, BitSet::B11,
            BitSet::B12, BitSet::B13, BitSet::B14, BitSet::B15,
            BitSet::B16, BitSet::B17, BitSet::B18, BitSet::B19,
            BitSet::B20, BitSet::B21, BitSet::B22, BitSet::B23,
            BitSet::B24, BitSet::B25, BitSet::B26, BitSet::B27,
            BitSet::B28, BitSet::B29, BitSet::B30, BitSet::B31,
        ];
        $max32bit = 0b1111111111111111111111111111111; // 2^31 − 1

        $bitSet = BitSet::set($b32a);
        static::assertSame($max32bit, $bitSet);
        static::assertSame(0b0, BitSet::remove($bitSet, $b32a));
    }
}


tests/unit/BitSet64BitTest.php
<?php
declare(strict_types=1);

namespace Pwm\BitSet;

use PHPUnit\Framework\TestCase;

class BitSet64BitTest extends TestCase
{
    /**
     * @test
     */
    public function setting_all_64_bits_result_in_the_max_signed_integer_on_64_bit_systems(): void
    {
        $b64a = [
            BitSet::B0,  BitSet::B1,  BitSet::B2,  BitSet::B3,
            BitSet::B4,  BitSet::B5,  BitSet::B6,  BitSet::B7,
            BitSet::B8,  BitSet::B9,  BitSet::B10, BitSet::B11,
            BitSet::B12, BitSet::B13, BitSet::B14, BitSet::B15,
            BitSet::B16, BitSet::B17, BitSet::B18, BitSet::B19,
            BitSet::B20, BitSet::B21, BitSet::B22, BitSet::B23,
            BitSet::B24, BitSet::B25, BitSet::B26, BitSet::B27,
            BitSet::B28, BitSet::B29, BitSet::B30, BitSet::B31,
            BitSet::B32, BitSet::B33, BitSet::B34, BitSet::B35,
            BitSet::B36, BitSet::B37, BitSet::B38, BitSet::B39,
            BitSet::B40, BitSet::B41, BitSet::B42, BitSet::B43,
            BitSet::B44, BitSet::B45, BitSet::B46, BitSet::B47,
            BitSet::B48, BitSet::B49, BitSet::B50, BitSet::B51,
            BitSet::B52, BitSet::B53, BitSet::B54, BitSet::B55,
            BitSet::B56, BitSet::B57, BitSet::B58, BitSet::B59,
            BitSet::B60, BitSet::B61, BitSet::B62, BitSet::B63,
        ];
        $max64bit = 0b111111111111111111111111111111111111111111111111111111111111111; // 2^63 − 1

        $bitSet = BitSet::set($b64a);
        static::assertSame($max64bit, $bitSet);
        static::assertSame(0b0, BitSet::remove($bitSet, $b64a));
    }
}


src/BitSet.php
<?php
declare(strict_types=1);

namespace Pwm\BitSet;

use InvalidArgumentException;

final class BitSet
{
    const EMPTY = self::B0;

    const B0  = 0b0; // 0
    const B1  = 0b1; // 2^0
    const B2  = 0b10; // 2^1
    const B3  = 0b100; // 2^2
    const B4  = 0b1000; // 2^3
    const B5  = 0b10000; // 2^4
    const B6  = 0b100000; // 2^5
    const B7  = 0b1000000; // 2^6
    const B8  = 0b10000000; // 2^7
    const B9  = 0b100000000; // 2^8
    const B10 = 0b1000000000; // 2^9
    const B11 = 0b10000000000; // 2^10
    const B12 = 0b100000000000; // 2^11
    const B13 = 0b1000000000000; // 2^12
    const B14 = 0b10000000000000; // 2^13
    const B15 = 0b100000000000000; // 2^14
    const B16 = 0b1000000000000000; // 2^15
    const B17 = 0b10000000000000000; // 2^16
    const B18 = 0b100000000000000000; // 2^17
    const B19 = 0b1000000000000000000; // 2^18
    const B20 = 0b10000000000000000000; // 2^19
    const B21 = 0b100000000000000000000; // 2^20
    const B22 = 0b1000000000000000000000; // 2^21
    const B23 = 0b10000000000000000000000; // 2^22
    const B24 = 0b100000000000000000000000; // 2^23
    const B25 = 0b1000000000000000000000000; // 2^24
    const B26 = 0b10000000000000000000000000; // 2^25
    const B27 = 0b100000000000000000000000000; // 2^26
    const B28 = 0b1000000000000000000000000000; // 2^27
    const B29 = 0b10000000000000000000000000000; // 2^28
    const B30 = 0b100000000000000000000000000000; // 2^29
    const B31 = 0b1000000000000000000000000000000; // 2^30 - limit on 32 bit systems
    const B32 = 0b10000000000000000000000000000000; // 2^31
    const B33 = 0b100000000000000000000000000000000; // 2^32
    const B34 = 0b1000000000000000000000000000000000; // 2^33
    const B35 = 0b10000000000000000000000000000000000; // 2^34
    const B36 = 0b100000000000000000000000000000000000; // 2^35
    const B37 = 0b1000000000000000000000000000000000000; // 2^36
    const B38 = 0b10000000000000000000000000000000000000; // 2^37
    const B39 = 0b100000000000000000000000000000000000000; // 2^38
    const B40 = 0b1000000000000000000000000000000000000000; // 2^39
    const B41 = 0b10000000000000000000000000000000000000000; // 2^40
    const B42 = 0b100000000000000000000000000000000000000000; // 2^41
    const B43 = 0b1000000000000000000000000000000000000000000; // 2^42
    const B44 = 0b10000000000000000000000000000000000000000000; // 2^43
    const B45 = 0b100000000000000000000000000000000000000000000; // 2^44
    const B46 = 0b1000000000000000000000000000000000000000000000; // 2^45
    const B47 = 0b10000000000000000000000000000000000000000000000; // 2^46
    const B48 = 0b100000000000000000000000000000000000000000000000; // 2^47
    const B49 = 0b1000000000000000000000000000000000000000000000000; // 2^48
    const B50 = 0b10000000000000000000000000000000000000000000000000; // 2^49
    const B51 = 0b100000000000000000000000000000000000000000000000000; // 2^50
    const B52 = 0b1000000000000000000000000000000000000000000000000000; // 2^51
    const B53 = 0b10000000000000000000000000000000000000000000000000000; // 2^52
    const B54 = 0b100000000000000000000000000000000000000000000000000000; // 2^53
    const B55 = 0b1000000000000000000000000000000000000000000000000000000; // 2^54
    const B56 = 0b10000000000000000000000000000000000000000000000000000000; // 2^55
    const B57 = 0b100000000000000000000000000000000000000000000000000000000; // 2^56
    const B58 = 0b1000000000000000000000000000000000000000000000000000000000; // 2^57
    const B59 = 0b10000000000000000000000000000000000000000000000000000000000; // 2^58
    const B60 = 0b100000000000000000000000000000000000000000000000000000000000; // 2^59
    const B61 = 0b1000000000000000000000000000000000000000000000000000000000000; // 2^60
    const B62 = 0b10000000000000000000000000000000000000000000000000000000000000; // 2^61
    const B63 = 0b100000000000000000000000000000000000000000000000000000000000000; // 2^62 - limit on 64 bit systems

    public static function get(int $bitSet): array
    {
        $values = [];
        $value = self::B1;
        while ($bitSet !== 0) {
            if (($bitSet & $value) > self::B0) {
                $values[] = $value;
                $bitSet &= ~$value;
            }
            $value <<= self::B1;
        }
        return $values;
    }

    public static function set(array $values): int
    {
        self::enforceSingleValues($values);
        $bitSet = self::B0;
        if (count($values) > 0) {
            foreach ($values as $value) {
                $bitSet |= $value;
            }
        }
        return $bitSet;
    }

    public static function add(int $bitSet, array $values): int
    {
        self::enforceSingleValues($values);
        return $bitSet | self::set($values);
    }

    public static function remove(int $bitSet, array $values): int
    {
        self::enforceSingleValues($values);
        return $bitSet & ~self::set($values);
    }

    public static function has(int $bitSet, int $value): bool
    {
        if ($value === self::B0) {
            return true;
        }
        self::enforceSingleValue($value);
        return ($bitSet & $value) > self::B0;
    }

    private static function enforceSingleValues(array $values): void
    {
        foreach ($values as $value) {
            self::enforceSingleValue($value);
        }
    }

    private static function enforceSingleValue(int $value): void
    {
        if (count(self::get($value)) > 1) {
            throw new InvalidArgumentException('Individual bit values must be equal to powers of 2');
        }
    }
}


example/User.php
<?php
declare(strict_types=1);

require_once __DIR__.'/../src/BitSet.php';

use Pwm\BitSet\BitSet;

class User
{
    private $groups = BitSet::EMPTY;

    public function getGroups(): array
    {
        return Group::toGroups(BitSet::get($this->groups));
    }

    public function setGroups(array $groups): void
    {
        $this->groups = BitSet::set(Group::fromGroups($groups));
    }

    public function addGroups(array $groups): void
    {
        $this->groups = BitSet::add($this->groups, Group::fromGroups($groups));
    }

    public function removeGroups(array $groups): void
    {
        $this->groups = BitSet::remove($this->groups, Group::fromGroups($groups));
    }

    public function hasGroup(string $group): bool
    {
        $a = Group::fromGroups([$group]);
        return isset($a[0])
            ? BitSet::has($this->groups, $a[0])
            : false;
    }
}


example/user.php
<?php
declare(strict_types=1);

require __DIR__.'/vendor/autoload.php';

use Pwm\BitSet\BitSet;

class User
{
    const G1 = 'Group 1';
    const G2 = 'Group 2';
    const G3 = 'Group 3';
    const G4 = 'Group 4';
    const G5 = 'Group 5';

    const GROUPS = [
        self::G1 => 0b1,
        self::G2 => 0b10,
        self::G3 => 0b100,
        self::G4 => 0b1000,
        self::G5 => 0b10000,
        // 32 bit max: 0b1000000000000000000000000000000
        // 64 bit max: 0b100000000000000000000000000000000000000000000000000000000000000
    ];

    private $groups = 0b0;

    public function getGroups(): array
    {
        return self::groupsFromBits(BitSet::get($this->groups));
    }

    public function setGroups(array $groups): void
    {
        $this->groups = BitSet::set(self::bitsFromGroups($groups));
    }

    public function addGroups(array $groups): void
    {
        $this->groups = BitSet::add($this->groups, self::bitsFromGroups($groups));
    }

    public function removeGroups(array $groups): void
    {
        $this->groups = BitSet::remove($this->groups, self::bitsFromGroups($groups));
    }

    public function hasGroup(string $group): bool
    {
        return BitSet::has($this->groups, self::bitsFromGroups([$group])[0]);
    }

    private static function bitsFromGroups($groups): array
    {
        return array_values(array_intersect_key(self::GROUPS, array_flip($groups)));
    }

    private static function groupsFromBits($groups): array
    {
        return array_values(array_intersect_key(array_flip(self::GROUPS), array_flip($groups)));
    }
}

$user = new User();
assert($user->getGroups() === []);
assert($user->hasGroup(User::G2) === false);

$user->setGroups([User::G1, User::G2]);
assert($user->getGroups() === [User::G1, User::G2]);
assert($user->hasGroup(User::G2) === true);

$user->addGroups([User::G4, User::G5]);
assert($user->getGroups() === [User::G1, User::G2, User::G4, User::G5]);
assert($user->hasGroup(User::G2) === true);
assert($user->hasGroup(User::G4) === true);

$user->removeGroups([User::G1, User::G4]);
assert($user->getGroups() === [User::G2, User::G5]);
assert($user->hasGroup(User::G2) === true);
assert($user->hasGroup(User::G4) === false);
